### Scaler classes

# Class 1

## Intro

In [1]:
!pip install numpy

Above command downloaded the file on the local system now we can use import term to import numpy

In [3]:
import numpy as np

-------Let us see the magic of numpy. we will use a python list a = [1,2,3,4,5]

In [5]:
a = [1,2,3,4,5]
a

[1, 2, 3, 4, 5]

a here is a list

In [6]:
b= [i**2 for i in a]
b

[1, 4, 9, 16, 25]

In [7]:
c = []
for i in a:
    c.append(i ** 2)
c

[1, 4, 9, 16, 25]

In [8]:
d = np.array([1,2,3,4,5])
d

array([1, 2, 3, 4, 5])

whenever we see array written, it is a numpy array.

In [9]:
d**2

array([ 1,  4,  9, 16, 25])

In [10]:
d + 2

array([3, 4, 5, 6, 7])

In [11]:
d*10

array([10, 20, 30, 40, 50])

In [12]:
d-2

array([-1,  0,  1,  2,  3])

In [19]:
r = range(10000)
r

range(0, 10000)

In [49]:
[i**2 for i in r]


AttributeError: 'list' object has no attribute 'head'

In [21]:
t = np.array(r)
t

array([   0,    1,    2, ..., 9997, 9998, 9999])

In [22]:
t**2

array([       0,        1,        4, ..., 99940009, 99960004, 99980001])

magic functions: are function provided by jupyter notebook. we can read about in documentation by the name built in magic commands

In [25]:
#to find time take by any code
%timeit [i**2 for i in r]

534 µs ± 3.28 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [26]:
%timeit t**2

3.47 µs ± 20.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


time taken by numpy array is 1000 times faster than anormal python for loop.

**Functions of numpy:**  
1.Homogeneous  
2.Fast  
3.simple syntax  
4.Functions- while using numpy we can use functions created by anyone else. but we don't have this advantage with python.  

## From a list- array(), shape, ndim

### 1. Shape

In [27]:
a=np.array([1,2,3,4,5,6])
a

array([1, 2, 3, 4, 5, 6])

In [28]:
a.shape

(6,)

**It shows a tupple with 6 elements.  
(6,) and (6) are different. first one is a tupple and second one is an integer value.**

In [29]:
a.shape[0]

6

In [30]:
len(a)

6

In [31]:
a.size

6

**answers to these two are same in a 1 d array but they will differ in a 2D array.**

### 2. ndim

It gives us the number of dimensions.  
also, number of brackets= number of dimensions.

## From a range and a stepsize - arange()

In [33]:
for i in range(1,10,2):
    print(i)

1
3
5
7
9


In [34]:
for i in range(1,10,0.5):
    print(i)

TypeError: 'float' object cannot be interpreted as an integer

It will give an error- they don't accept float steps.  
**Now arange comes into the picture, it accepts float steps. It avoids writing 0 after decimel to save space.  
example: it won't write 1.0, instead it will write 1.**

In [35]:
np.arange(1,10,0.5)

array([1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. , 5.5, 6. , 6.5, 7. ,
       7.5, 8. , 8.5, 9. , 9.5])

## type( ) ndarray 

In [36]:
a = np.array([1,2,3,4,5])
a

array([1, 2, 3, 4, 5])

In [37]:
type(a)

numpy.ndarray

**We will get ndarray. Ndarray means n-dimensional array. To know the type of data type it is storing, we will use _"DTYPE"_ key word**

In [38]:
a.dtype

dtype('int64')

In [42]:
a = np.array([1.0,2.0,3.0,4.0,5.0])
print(a)
print(type(a))
print(a.dtype)

[1. 2. 3. 4. 5.]
<class 'numpy.ndarray'>
float64


In [43]:
a = np.array([True, False])
print(a)
print(type(a))
print(a.dtype)

[ True False]
<class 'numpy.ndarray'>
bool


In [44]:
a = np.array([True, False, 2])
print(a)
print(type(a))
print(a.dtype)

[1 0 2]
<class 'numpy.ndarray'>
int64


**This will not throw an error, it will infact convert boolean into integer. Upcast boolean as integer**

In [45]:
a = np.array([True, False, 2,3.0])
print(a)
print(type(a))
print(a.dtype)

[1. 0. 2. 3.]
<class 'numpy.ndarray'>
float64


**This will convert boolean and integer into float type.**

In [46]:
a = np.array([True, False, 2,3.0, "scaler"])
print(a)
print(type(a))
print(a.dtype)

['True' 'False' '2' '3.0' 'scaler']
<class 'numpy.ndarray'>
<U32


**This will upcast boolean, integer and float type into strings.Everthing converst to string.  
U32 is Little Endian, U is Unicode and 32 is the limit for character.**

In [6]:
a = np.array(['Scaler', 'Academy', 'Learner'], dtype='U2')
print(a)
print(type(a))
print(a.dtype)

['Sc' 'Ac' 'Le']
<class 'numpy.ndarray'>
<U2


**So here we only see 2 characters only, as we declared U2.**

# Class 2

## Indexing, slicing, Masking on 1D

### Indexing

In [14]:
a=np.arange(1,9)
print(a)
print(a[3])
print(a[-5])

[1 2 3 4 5 6 7 8]
4
4


**Indexing a single element is very easy, we can directly write the index numbers. But to get multiple elements where there is no corelation and we cannot do slicing. We can simply create a list of indexes we want, and pass the list inside the 'a'.**

In [15]:
[2,4,5]

[2, 4, 5]

In [16]:
a[[2,4,5]]

array([3, 5, 6])

**Let's change the order**

In [18]:
a[[5,4,2]]

array([6, 5, 3])

In [24]:
print(a[[-6,-4,-3]])
print(a[[-6,4,-3]])

[3 5 6]
[3 5 6]


**This means we can pass mix and match of positive and negative indices.**

### Slicing

In [38]:
a=np.arange(11,19)
a

array([11, 12, 13, 14, 15, 16, 17, 18])

In [30]:
print(a[2:7:1])
print(a[2::1])
print(a[-6:-1:1])

[13 14 15 16 17]
[13 14 15 16 17 18]
[13 14 15 16 17]


In [31]:
a[4:2:1]

array([], dtype=int64)

**In numpy we don't get any errors we will just get an empty array.  
Implying that there is no element looking forward.**  

**It shows datatype because it is an empty array, and every time we create an array it stores some data type. So this output wants to say that we _"we have an empty array which stores integer values"_**

In [32]:
a[4:2:-1]

array([15, 14])

In [33]:
a[-4:-2:-1]

array([], dtype=int64)

In [35]:
a[1:-2:2]

array([12, 14, 16])

In [36]:
a[6:-7:-3]

array([17, 14])

In [37]:
a[5::-2]

array([16, 14, 12])

In [39]:
a[:3:-3]

array([18, 15])

**Solve it manually**

In [40]:
a[::-1]

array([18, 17, 16, 15, 14, 13, 12, 11])

**This gives reverse of the list.**

### Masking (Fancy Indexing)

In [41]:
a

array([11, 12, 13, 14, 15, 16, 17, 18])

In [42]:
a+10

array([21, 22, 23, 24, 25, 26, 27, 28])

**A stays same,this is not doing in place changes.**

In [43]:
a

array([11, 12, 13, 14, 15, 16, 17, 18])

In [44]:
a<15

array([ True,  True,  True,  True, False, False, False, False])

In [46]:
a[a<15]

array([11, 12, 13, 14])

In [47]:
a[a%2==0]

array([12, 14, 16, 18])

**In the next business case we will see how we can use these Masking for _"String"_ as well.**

In [48]:
a[~a%2==0]

array([11, 13, 15, 17])

**_Tilda_ is used to get opposite/not values**. 


**Masking is faster than filter and lambda. Masking is the fastest. Internal working is vectorisation and broadcasting**

In [50]:
mask1 = a%2==0
mask2 = a%5==0
mask1

array([False,  True, False,  True, False,  True, False,  True])

In [51]:
mask2

array([False, False, False, False,  True, False, False, False])

In [52]:
a

array([11, 12, 13, 14, 15, 16, 17, 18])

**In python we did and and or, but it doesnt work here as they compare the entire object. We are doing a bit wise comparison here.  
We are comparinf individual comparison of respective False's and Trues  
![bitwise comparison](/Users/aparnabiswas/Downloads/Python libraries-Jupyter)

In [53]:
a[mask1 | mask2]

array([12, 14, 15, 16, 18])

In [54]:
a[mask1 and mask2]

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [55]:
a[mask1 & mask2]

array([], dtype=int64)

Everything in python is an object when we do **and** operation entire array of mask1/mask2 of [True, False,....], entire array will be compared to each other in whole. BUt we need to do Bit-wise comparison. Hence **"&"** is used.

In [56]:
a[a%2==0 | a%3==0]

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

**Error**:
as there is certain order of precendence with all the operators.  
Hence, use Brackets.

In [57]:
a[(a%2==0) | (a%3==0)]

array([12, 14, 15, 16, 18])

## Operation on arrays

#### np.any, np.all

In [59]:
a=np.array([1,2,3,4,5])
b=np.array ( [6, 7, 8, 9, 10])
c=np.array([6,7,8,9,10,11])

In [60]:
a+10

array([11, 12, 13, 14, 15])

In [61]:
a<10

array([ True,  True,  True,  True,  True])

In [62]:
a+b

array([ 7,  9, 11, 13, 15])

In [64]:
#a+c #error as the shape is not same

In [66]:
a<b
#everything in "a" & "b" will be compared individually.

array([ True,  True,  True,  True,  True])

In [67]:
a>b

array([False, False, False, False, False])

In [68]:
a==b

array([False, False, False, False, False])

**np.any** - even if one value is true  
**np.all** - When all value is true

In [70]:
np.all(a<b)

True

In [71]:
a=np.array ([1,2,3,4,11])
b=np.array ( [6,7,8,9, 10])

In [72]:
#now
np.all(a<b)

False

In [73]:
np.any(a<b)

True

### U Functions-

Universal function- aggregate functions

#### np.max

#### np.min

#### np.mean

#### np.median

#### np.sum

## NPS Biz case

We will use **np.loadtxt** to load the dataset it is in"txt" form. But generally we will use pd.read_csv to load the file.

In [76]:
data = np.loadtxt("/Users/aparnabiswas/Downloads/Python_libraries-Jupyter/Numpy/NPS_Biz_case_class2/survey.txt") 
data
#storing this inside data

array([ 7., 10.,  5., ...,  5.,  9., 10.])

Formula of **NPS**:  
%p - %d

In [77]:
len(data)

1167

**to find total number of feedbacks. Since this is 1D, we can just use  
"len" command
Shape[0]
size**


In [78]:
data.shape[0]

1167

In [80]:
total=data.size
total

1167

In [81]:
#detractors
data[data<7]

array([5., 4., 4., 5., 1., 5., 5., 1., 4., 5., 4., 4., 4., 5., 1., 4., 1.,
       4., 1., 5., 5., 1., 1., 4., 1., 5., 4., 1., 1., 4., 1., 5., 1., 4.,
       4., 1., 1., 1., 1., 1., 1., 1., 4., 1., 1., 5., 5., 5., 4., 4., 1.,
       4., 1., 4., 1., 5., 1., 1., 5., 4., 4., 4., 4., 1., 4., 5., 4., 4.,
       1., 1., 5., 5., 1., 5., 1., 5., 5., 4., 5., 4., 1., 1., 1., 1., 4.,
       1., 4., 4., 5., 4., 1., 1., 1., 1., 5., 4., 5., 5., 4., 1., 5., 1.,
       4., 4., 1., 1., 1., 4., 4., 5., 5., 4., 5., 5., 5., 1., 4., 1., 5.,
       5., 1., 5., 1., 1., 5., 5., 4., 4., 1., 4., 4., 4., 1., 1., 4., 4.,
       4., 5., 5., 1., 1., 4., 1., 4., 1., 4., 1., 4., 1., 4., 1., 5., 4.,
       4., 5., 1., 4., 5., 5., 5., 1., 5., 4., 1., 1., 5., 5., 5., 4., 5.,
       4., 4., 1., 4., 4., 4., 4., 5., 1., 5., 5., 1., 4., 4., 5., 1., 1.,
       4., 5., 5., 5., 1., 4., 5., 5., 4., 1., 5., 5., 5., 1., 1., 5., 5.,
       1., 1., 1., 4., 5., 5., 4., 4., 4., 5., 1., 4., 1., 4., 5., 4., 5.,
       5., 1., 5., 1., 5.

In [87]:
#to find how many detractors

detractors = data[data<7].shape[0]
detractors

332

In [88]:
#promoters

promoters = data[data>=9].shape[0]
promoters

609

In [90]:
#passives

passives = data[(data==7)|(data==8)].shape[0]
passives

226

In [91]:
perc_promoters = promoters/total
perc_promoters

0.5218508997429306

In [92]:
perc_detractors = detractors/total
perc_detractors

0.28449014567266495

In [94]:
NPS = (perc_promoters-perc_detractors)*100
NPS

23.73607540702657

## 2D Numpy array

In [95]:
a=np.array([[1,2,3],[4,5,6]])
a

array([[1, 2, 3],
       [4, 5, 6]])

In [96]:
a.ndim  #tells us number of dimensions

2

In [97]:
a.shape #it tells us number of elements along a particular dimension

(2, 3)

**We can have objects - list, tupples, dictionaries, inside numpy arrays. But we don't use them.**
example of list- (1,1.0,"string")  

Reasons why we don't use them-  
these lists and dictionaries are mutable. We can increase and decrease the size. Hence, we are losing out on the capabilities of fast and efficient numpy methods. BEcause it will be same as a python list.

In [99]:
a.size #total number of eelemts inside the numpy array.

6

In [100]:
len(a)

2

In [101]:
a=np.arange(1,13)
a

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [102]:
a.ndim

1

In [103]:
a.shape

(12,)

**Reshape**

In [104]:
a.reshape((3,4))

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [105]:
a.reshape((1,12))

array([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]])

In [106]:
a.reshape((12,1))

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12]])

In [107]:
a #remains unaffected

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [108]:
a.reshape((5,2))

ValueError: cannot reshape array of size 12 into shape (5,2)

**Concept of negative value in reshape**

In [113]:
a.reshape((-1,2))  #numpy decides,even if we give just one unknown dimension

array([[ 1,  2],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10],
       [11, 12]])

In [112]:
a.reshape((2,-1))#numpy decides,even if we give just one unknown dimension

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12]])

In [111]:
a.reshape(-1,-1)

ValueError: can only specify one unknown dimension

In [114]:
a.reshape((-1,)) #num of dimensions- 1D

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [115]:
a.reshape((-1,5))

ValueError: cannot reshape array of size 12 into shape (5)

In [116]:
a.reshape((,6)) #

SyntaxError: invalid syntax (183478571.py, line 1)

**Here (,6) is not a tuple. So Invalid syntax error comes.**

In [117]:
a.reshape((3,2,2))

array([[[ 1,  2],
        [ 3,  4]],

       [[ 5,  6],
        [ 7,  8]],

       [[ 9, 10],
        [11, 12]]])

In [118]:
a.reshape((3,2,-1))

array([[[ 1,  2],
        [ 3,  4]],

       [[ 5,  6],
        [ 7,  8]],

       [[ 9, 10],
        [11, 12]]])

In [119]:
a.reshape((3,-1,-1))

ValueError: can only specify one unknown dimension

**As numpy can handle only one unknown variable.**  
**We will with tuples,dictionaries in Panda.**

# Class 3

## 2D arrays(Matrices)

### Transpose- a.transpose( ), a.T

Transpose is converting rows into columns and columns into rows.

In [120]:
a = np.arange(1,13)
a =a.reshape(3,4)
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [121]:
a.transpose()

array([[ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11],
       [ 4,  8, 12]])

In [122]:
a.T

array([[ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11],
       [ 4,  8, 12]])

In [124]:
b=np.arange(1,13)
c=b.reshape((1,12))

In [125]:
b

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [126]:
c

array([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]])

In [128]:
b.shape

(12,)

In [129]:
c.shape

(1, 12)

In [130]:
c.T

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12]])

In [131]:
b.T

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

It stays same as it has only 1 dimension. And we cannot move into another direction.

### Flatten

**Any "n" dimensional array can be converted to a 1 Dimensional array**

In [132]:
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [135]:
a.flatten()  #doesn't change the a array

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

### Indexing, Slicing and Masking on 2D

#### Indexing

In [134]:
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

**Method 1**

In [138]:
a[1][1]  #don't use this

6

**Method 2**

In [137]:
a[1,1] #right way of indexing a[Rows,columns]


6

In [139]:
a[-1,-2]

11

In [140]:
a[2,-2]

11

In [142]:
a[[1,2,2],[1,2,3]]   #[row indexes], [column indexes]

array([ 6, 11, 12])

In [143]:
a[[0,2,2],[2,3,0]]

array([ 3, 12,  9])

In [144]:
a[[-2,-3,-1,-2],[-3,-1,-3,-1]]

array([ 6,  4, 10,  8])

#### Slicing

In [145]:
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [146]:
a[::,::]

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [147]:
a[:2:1,::]

array([[1, 2, 3, 4],
       [5, 6, 7, 8]])

In [148]:
a[:2:1,1::2]

array([[2, 4],
       [6, 8]])

In [149]:
 a[1::-1,2::1]

array([[7, 8],
       [3, 4]])

In [150]:
a[1::-1,2::-1]

array([[7, 6, 5],
       [3, 2, 1]])

In [153]:
a[0:2:2,-2::-2]

array([[3, 1]])

In [154]:
a[0:2:2,-2:-1:-2]

array([], shape=(1, 0), dtype=int64)

In [156]:
a[::-2,:-4:-3]

array([[12],
       [ 4]])

#### Masking

In [169]:
a=np.array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [158]:
a+10

array([[11, 12, 13, 14],
       [15, 16, 17, 18],
       [19, 20, 21, 22]])

In [159]:
a<10

array([[ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True, False, False, False]])

In [160]:
a[a<10]

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

**We get an 1 D array. Why?  
Creators thought if we keep maksing output in 1 D, we can get answers in a fixed format. It wouldn't change as per the dimension, Cz we don't know how many elements in output we will get.**

In [161]:
a[(a%2==0)|(a%5==0)]

array([ 2,  4,  5,  6,  8, 10, 12])

### Universal Functions

np.mean  
np.max  
np.sum. 
etc

#### Axis Argument

Any Changw/rearrangement in 2 D numpy arrays, if one is moving  
vertical axis- value of axis is 0.  
and if we are moving in horizontal direction value of axis will be 1

In [170]:
a = np.array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [172]:
np.sum(a,axis=0)

array([15, 18, 21, 24])

In [173]:
np.sum(a,axis=1)

array([10, 26, 42])

In [174]:
np.sum(a,axis=-1)

array([10, 26, 42])

In [175]:
np.sum(a,axis=-2)

array([15, 18, 21, 24])

In [176]:
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [177]:
np.min(a,axis=-2)

array([1, 2, 3, 4])

In [178]:
np.min(a,axis=1)

array([1, 5, 9])

In [179]:
np.min(a,axis = -1)

array([1, 5, 9])

#### Logical Operation

In [190]:
a

array([[ 10,   2,  99],
       [100,   4,  21],
       [ 32,   5,  12],
       [  9,  57,  45]])

In [191]:
a<10

array([[False,  True, False],
       [False,  True, False],
       [False,  True, False],
       [ True, False, False]])

In [192]:
a[a<10]

array([2, 4, 5, 9])

In [193]:
b=np.arange(1,13).reshape((4,3))
b

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [194]:
a<b

array([[False, False, False],
       [False,  True, False],
       [False,  True, False],
       [ True, False, False]])

In [195]:
np.all(a<b)

False

In [196]:
np.any(a<b)

True

In [ ]:
#1D np array

In [209]:
a=np.array([2,7,3,9,4,1,6,7])
a

array([2, 7, 3, 9, 4, 1, 6, 7])

**#Method 1**

In [203]:
print(np.sort(a))
print(a)

[1 2 3 4 6 7 7 9]
[2 7 3 9 4 1 6 7]


**#Method 2:**  **This is infact changing the entire array and sorting it**

In [204]:
a.sort()
a

array([1, 2, 3, 4, 6, 7, 7, 9])

In [207]:
np.max(a)

9

In [206]:
a.max()

9

#### Sorting function - `sort( ), argsort( )`

In [180]:
a= np.array([[10, 2, 99], 
        [100, 4, 21],
        [ 32,  5, 12],
        [ 9,  57, 45]])

In [183]:
a

array([[ 10,   2,  99],
       [100,   4,  21],
       [ 32,   5,  12],
       [  9,  57,  45]])

In [184]:
np.sort(a,axis=0)

array([[  9,   2,  12],
       [ 10,   4,  21],
       [ 32,   5,  45],
       [100,  57,  99]])

In [185]:
np.sort(a,axis=-2)

array([[  9,   2,  12],
       [ 10,   4,  21],
       [ 32,   5,  45],
       [100,  57,  99]])

In [187]:
np.sort(a,axis=1)

array([[  2,  10,  99],
       [  4,  21, 100],
       [  5,  12,  32],
       [  9,  45,  57]])

In [188]:
np.sort(a,axis=-1)

array([[  2,  10,  99],
       [  4,  21, 100],
       [  5,  12,  32],
       [  9,  45,  57]])

In [189]:
np.sort(a) ##default value is -1.

array([[  2,  10,  99],
       [  4,  21, 100],
       [  5,  12,  32],
       [  9,  45,  57]])

**`Default Value of sorting- numpy.sort`**   
  
**numpy.sort (a, axis=-1, kind=None, order=None)  
Return a sorted copy of an array.**  

In most functions default value is -1. So check function to function.  
In sort function default value is **-1**

##### `np.argsort( )` - Gives indexes of sorted elements

In [210]:
np.argsort(a) #

array([5, 0, 2, 4, 6, 1, 7, 3])

In [212]:
np.argmax(a) #returns index of maximum value/argument

3

In [213]:
np.argmin(a) #returns index of minimum value/argument

5

#### `np.unique( )`

In [222]:
data

array([ 7., 10.,  5., ...,  5.,  9., 10.])

**Finds unique value**

In [223]:
np.unique(data) 

array([ 1.,  4.,  5.,  7.,  8.,  9., 10.])

**To find frequency of Unique values**

In [224]:
np.unique(data,return_counts=True)

(array([ 1.,  4.,  5.,  7.,  8.,  9., 10.]),
 array([117, 112, 103, 123, 103, 394, 215]))

**In the above output, first array is all the unique values and the second array has the frequency of the unique values.**  

Ex: Those who gave 1 rating are 117 in number.   
Those who gave 4. rating they are 112 in number.

#### `np.where( )`

In [225]:
a = np.array([2,1,6,4,7,3,8,9,11])
a

array([ 2,  1,  6,  4,  7,  3,  8,  9, 11])

In [226]:
a<6

array([ True,  True, False,  True, False,  True, False, False, False])

In [227]:
a[a<6]

array([2, 1, 4, 3])

In [228]:
np.where(a<6)

(array([0, 1, 3, 5]),)

**We receive the indexes of the values lesser than 6.**  

Why is there a comma? cz it is a tuple. To store a single element in Tuple we have to write a comma.

In [229]:
np.where(a<6,-1,+1)

array([-1, -1,  1, -1,  1, -1,  1,  1,  1])

In above code- we replaced values which are <6 with -1 and values which are >6 with +1

In [230]:
np.where(a<6,-1,a)

array([-1, -1,  6, -1,  7, -1,  8,  9, 11])

In above code- we replaced values which are <6 with -1 and values which are >6 are kept as it is.

### Use case- Fitness data analysis

In [232]:
data = np.loadtxt("/Users/aparnabiswas/Downloads/Python_libraries-Jupyter/Numpy/04_Numpy_4/fit.txt",dtype = "str")
data

array([['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
       ['07-10-2017', '6041', 'Sad', '197', '8', 'Inactive'],
       ['08-10-2017', '25', 'Sad', '0', '5', 'Inactive'],
       ['09-10-2017', '5461', 'Sad', '174', '4', 'Inactive'],
       ['10-10-2017', '6915', 'Neutral', '223', '5', 'Active'],
       ['11-10-2017', '4545', 'Sad', '149', '6', 'Inactive'],
       ['12-10-2017', '4340', 'Sad', '140', '6', 'Inactive'],
       ['13-10-2017', '1230', 'Sad', '38', '7', 'Inactive'],
       ['14-10-2017', '61', 'Sad', '1', '5', 'Inactive'],
       ['15-10-2017', '1258', 'Sad', '40', '6', 'Inactive'],
       ['16-10-2017', '3148', 'Sad', '101', '8', 'Inactive'],
       ['17-10-2017', '4687', 'Sad', '152', '5', 'Inactive'],
       ['18-10-2017', '4732', 'Happy', '150', '6', 'Active'],
       ['19-10-2017', '3519', 'Sad', '113', '7', 'Inactive'],
       ['20-10-2017', '1580', 'Sad', '49', '5', 'Inactive'],
       ['21-10-2017', '2822', 'Sad', '86', '6', 'Inactive'],
       ['22-10

**Numpys don't deal with date format, hence we will use Pandas for it later.  
and for now we have converted everything into string.**

In [233]:
step_counts

NameError: name 'step_counts' is not defined

In [234]:
date = data[::,0]
step_counts = data[::,1]
mood= data[::,2]
calories_burnt= data[::,3]
hours_of_sleep = data[::,4]
activity = data[::,5]

In [235]:
date

array(['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
       '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
       '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
       '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
       '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
       '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
       '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
       '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
       '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
       '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
       '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
       '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
       '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
       '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
       '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
       '05-12-2017', '06-12-2017', '07-12-2017', '08-12

In [236]:
step_counts

array(['5464', '6041', '25', '5461', '6915', '4545', '4340', '1230', '61',
       '1258', '3148', '4687', '4732', '3519', '1580', '2822', '181',
       '3158', '4383', '3881', '4037', '202', '292', '330', '2209',
       '4550', '4435', '4779', '1831', '2255', '539', '5464', '6041',
       '4068', '4683', '4033', '6314', '614', '3149', '4005', '4880',
       '4136', '705', '570', '269', '4275', '5999', '4421', '6930',
       '5195', '546', '493', '995', '1163', '6676', '3608', '774', '1421',
       '4064', '2725', '5934', '1867', '3721', '2374', '2909', '1648',
       '799', '7102', '3941', '7422', '437', '1231', '1696', '4921',
       '221', '6500', '3575', '4061', '651', '753', '518', '5537', '4108',
       '5376', '3066', '177', '36', '299', '1447', '2599', '702', '133',
       '153', '500', '2127', '2203'], dtype='<U10')

**Converting string back to integer**

In [239]:
step_counts = np.array(step_counts,dtype='int')
step_counts

array([5464, 6041,   25, 5461, 6915, 4545, 4340, 1230,   61, 1258, 3148,
       4687, 4732, 3519, 1580, 2822,  181, 3158, 4383, 3881, 4037,  202,
        292,  330, 2209, 4550, 4435, 4779, 1831, 2255,  539, 5464, 6041,
       4068, 4683, 4033, 6314,  614, 3149, 4005, 4880, 4136,  705,  570,
        269, 4275, 5999, 4421, 6930, 5195,  546,  493,  995, 1163, 6676,
       3608,  774, 1421, 4064, 2725, 5934, 1867, 3721, 2374, 2909, 1648,
        799, 7102, 3941, 7422,  437, 1231, 1696, 4921,  221, 6500, 3575,
       4061,  651,  753,  518, 5537, 4108, 5376, 3066,  177,   36,  299,
       1447, 2599,  702,  133,  153,  500, 2127, 2203])

In [240]:
calories_burnt  = np.array(calories_burnt,dtype='int')
calories_burnt

array([181, 197,   0, 174, 223, 149, 140,  38,   1,  40, 101, 152, 150,
       113,  49,  86,   6,  99, 143, 125, 129,   6,   9,  10,  72, 150,
       141, 156,  57,  72,  17, 181, 197, 131, 154, 137, 193,  19, 101,
       139, 164, 137,  22,  17,   9, 145, 192, 146, 234, 167,  16,  17,
        32,  35, 220, 116,  23,  44, 131,  86, 194,  60, 121,  76,  93,
        53,  25, 227, 125, 243,  14,  39,  55, 158,   7, 213, 116, 129,
        21,  28,  16, 180, 138, 176,  99,   5,   1,  10,  47,  84,  23,
         4,   0,   0,   0,   0])

In [241]:
hours_of_sleep = np.array(hours_of_sleep,dtype='int')
hours_of_sleep

array([5, 8, 5, 4, 5, 6, 6, 7, 5, 6, 8, 5, 6, 7, 5, 6, 8, 5, 4, 5, 6, 8,
       5, 6, 5, 8, 5, 4, 5, 4, 5, 4, 3, 2, 9, 5, 6, 4, 5, 8, 4, 5, 6, 5,
       6, 5, 6, 5, 6, 5, 6, 7, 6, 7, 6, 5, 6, 7, 8, 8, 7, 8, 5, 4, 3, 3,
       4, 5, 5, 5, 3, 4, 4, 5, 5, 5, 5, 5, 5, 4, 3, 4, 5, 5, 4, 5, 3, 3,
       3, 2, 3, 2, 8, 5, 5, 5])

In [242]:
#How many steps on an average the user is taking??

np.mean(step_counts)

2935.9375

In [244]:
#on how many days the user is sad?

np.unique(mood,return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([40, 27, 29]))

In [245]:
#on how many days the user is inactive/lazy?

np.unique(activity,return_counts=True)

(array(['Active', 'Inactive'], dtype='<U10'), array([42, 54]))

In [246]:
np.max(step_counts)

7422

In [247]:
np.argmax(step_counts)

69

**To find the date on which he walked maximum number of steps**

In [248]:
date[69]  

'14-12-2017'

In [249]:
date[np.argmax(step_counts)]

'14-12-2017'

In [251]:
mood=="Sad"

array([False,  True,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False,  True,
       False, False,  True,  True, False, False, False, False, False,
        True, False,  True, False, False,  True,  True,  True,  True,
       False, False, False, False,  True,  True,  True, False, False,
        True,  True, False, False, False, False])

In [253]:
step_counts[mood=="Sad"]

array([6041,   25, 5461, 4545, 4340, 1230,   61, 1258, 3148, 4687, 3519,
       1580, 2822,  181, 6676, 3721, 1648,  799, 1696,  221, 4061,  651,
        753,  518,  177,   36,  299,  702,  133])

In [254]:
np.mean(step_counts[mood=="Sad"])

2103.0689655172414

In [255]:
np.mean(step_counts[mood=="Neutral"])

3153.777777777778

In [256]:
np.mean(step_counts[mood=="Happy"])

3392.725

In [257]:
mood[step_counts>4000]

array(['Neutral', 'Sad', 'Sad', 'Neutral', 'Sad', 'Sad', 'Sad', 'Happy',
       'Neutral', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy',
       'Neutral', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Sad',
       'Happy', 'Happy', 'Neutral', 'Happy', 'Neutral', 'Neutral', 'Sad',
       'Happy', 'Neutral', 'Happy'], dtype='<U10')

In [258]:
np.unique(mood[step_counts>4000],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([22,  9,  7]))

In [260]:
np.unique(mood[step_counts<2000],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([13,  8, 18]))

In [261]:
np.unique(mood[hours_of_sleep<4],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([1, 5, 6]))

In [263]:
np.unique(mood[hours_of_sleep>5],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([20,  3, 10]))

**Recommendation to User**
- Sleep more
- Walk more

In [264]:
np.unique(mood[activity=="Active"],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([24, 14,  4]))

In [265]:
np.unique(mood[activity=="Inactive"],return_counts=True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([16, 13, 25]))

In [266]:
np.mean(hours_of_sleep[mood=="Happy"])

5.725

In [267]:
np.mean(hours_of_sleep[mood=="Sad"])

5.0344827586206895

### Vectorization

In [269]:
a= np.arange(1,11)
a

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

**What happens inside the array when we type a+10**  
Operation of +10 is happening inside individually on all the numbers.  
`2 things are happening-`  
`- Vectorisation`  
`- Broadcasting`

Read more in notes.

In [270]:
import math
math.log

<function math.log>

In [271]:
#to find log of all the values in a list- we use for loop.

l=[1,2,3,4,5,6,7,8,9,10]
l_e=[ ]
for i in l:
    l_e.append(math.log(i))
l_e

[0.0,
 0.6931471805599453,
 1.0986122886681098,
 1.3862943611198906,
 1.6094379124341003,
 1.791759469228055,
 1.9459101490553132,
 2.0794415416798357,
 2.1972245773362196,
 2.302585092994046]

In [272]:
#there are direct functions which gives exact result as above

np.log(a)

array([0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
       1.79175947, 1.94591015, 2.07944154, 2.19722458, 2.30258509])